### PR TITLE
tests: smoke test support for core22

### PR DESCRIPTION
**Pull request by Meulengracht**
This is a PR to get smoke tests up and running for core22.

There are some TODOs in this as some core22 snaps are not out of edge. And I had to update the snapd-testing-tools subtree to get the new commits for is-core22 etc.

### CODE REVIEW

File: tests/smoke/install/task.yaml [hunk 1]
```diff
@@ -44,9 +44,9 @@ execute: |
     snap changes | MATCH 'Install "test-snapd-sh" snap'
 
     echo "Ensure different bases work"
-    for base in core18 core20; do
-        # no core20 snap for i386
-        if [ "$base" = "core20" ] && [ "$(uname -m)" = i686 ]; then
+    for base in core18 core20 core22; do
+        # no core20 or core22 snap for i386
+        if { [ "$base" = "core20" ] || [ "$base" = "core22" ]; } && [ "$(uname -m)" = i686 ]; then
```
Comment (mvo5) [inline tests/smoke/install/task.yaml:49]: Hm, silly question, this `if {` pattern I only see rarely, usually I see `(` type brackets, why the `{` here? Is there an advantage?

**Review by mvo5 — APPROVED** [review]
Some questions/nitpicks inline but overall very nice, thank you!

File: tests/smoke/install/task.yaml (hunk 1 above)
Comment (Meulengracht) [reply tests/smoke/install/task.yaml:49]: Good question! I was hit over the head by shellcheck telling me this is the correct way to do this.

File: spread.yaml [hunk 2]
```diff
@@ -412,6 +412,18 @@ backends:
             - ubuntu-core-20-arm-32:
                   username: external
                   password: ubuntu
+            - ubuntu-core-22-64:
```
[the remark this answers is not in the document]
Comment (Meulengracht) [reply spread.yaml:415]: Really good point! I'll get it added

File: tests/lib/external/snapd-testing-tools/tools/repack-kernel [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+        fi
+    fi
+}
+
+get_kver() {
+    local kerneldir="$1"
+    (
+        cd "$kerneldir"
+        #shellcheck disable=SC2010
+        ls "config"-* | grep -Po 'config-\K.*'
```
Comment (alfonsosanchezbeato) [inline tests/lib/external/snapd-testing-tools/tools/repack-kernel:46]: I think something like:
`find "$kerneldir" -maxdepth 1 -name "config"-* | grep -Po 'config-\K.*'`
would avoid the subshell and the shellcheck warning.

File: tests/lib/external/snapd-testing-tools/tools/repack-kernel [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+        ubuntu-core-initramfs create-efi \
+                              --kernelver "$kver" \
+                              --initrd repacked-initrd \
+                              --kernel vmlinuz \
+                              --output repacked-kernel.efi
+
+        cp "repacked-kernel.efi-$kver" "$target/kernel/kernel.efi"
+
+        # XXX: needed?
+        chmod +x "$target/kernel/kernel.efi"
```
Comment (alfonsosanchezbeato) [inline tests/lib/external/snapd-testing-tools/tools/repack-kernel:138]: I can confirm it is not. At least for qemu :)

File: tests/lib/external/snapd-testing-tools/tools/repack-kernel [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+
+    if [ -z "$target" ]; then
+        echo "repack-kernel: missing target for cull-firmware"
+        exit 1
+    fi
+    (
+        # XXX: drop ~450MB+ of firmware which should not be needed in under qemu
+        # or the cloud system
+        rm -rf "$target"/kernel/firmware/*
+    )
```
Comment (alfonsosanchezbeato) [inline tests/lib/external/snapd-testing-tools/tools/repack-kernel:153]: Why the subshell here?

**Review by alfonsosanchezbeato — APPROVED** [review]
LGTM! I added some comments about the shell scripts, but then I remembered those are imported, iirc. Anyway, leaving those comments in case somebody wants to take a look at those in the imported repo.

File: tests/lib/external/snapd-testing-tools/tools/repack-kernel [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+
+setup() {
+    if [ "$UID" != "0" ]; then
+        echo "run as root (only this command)"
+        exit 1
+    fi
+
+    # carries ubuntu-core-initframfs
+    add-apt-repository ppa:snappy-dev/image -y
+    apt update -y
```
[the remark this answers is not in the document]
Comment (Meulengracht) [reply tests/lib/external/snapd-testing-tools/tools/repack-kernel:31]: Thanks for the review comments on repack-kernel, they are greatly appreciated! I'll handle them in a follow up PR as I can see Alfonso also has comments to the repack-kernel!

File: tests/lib/external/snapd-testing-tools/tools/repack-kernel (hunk 3 above)
Comment (Meulengracht) [reply tests/lib/external/snapd-testing-tools/tools/repack-kernel:46]: Awesome, thanks for the suggestion!

File: tests/lib/external/snapd-testing-tools/tools/repack-kernel (hunk 4 above)
Comment (Meulengracht) [reply tests/lib/external/snapd-testing-tools/tools/repack-kernel:138]: Thanks Alfonso!

File: tests/lib/external/snapd-testing-tools/tools/repack-kernel (hunk 5 above)
Comment (Meulengracht) [reply tests/lib/external/snapd-testing-tools/tools/repack-kernel:153]: I'll test to see if behaviour is fine without, then I'll remove it. Thanks for all your review comments in snapd-testing-tools, it's really appreciated! I'll create a followup PR with the modifications as I just need it go through snapd-testing-tools first!

File: tests/lib/external/snapd-testing-tools/tools/repack-kernel (hunk 3 above)
Comment (Meulengracht) [reply tests/lib/external/snapd-testing-tools/tools/repack-kernel:46]: I just tested this and your suggestion works just fine and avoids the subshell/shellcheck warning

File: spread.yaml [hunk 7]
```diff
@@ -412,6 +419,18 @@ backends:
...
+                  username: external
+                  password: ubuntu
+            - ubuntu-core-22-32:
```
Comment (sergiocazzolato) [inline spread.yaml:425]: do we need this one? I think it is not supported, we also should remove ubuntu-core-20-32

**Review by sergiocazzolato — APPROVED** [review]
Nice, thanks for adding core22, just a minor comment inline

File: spread.yaml (hunk 7 above)
Comment (Meulengracht) [reply spread.yaml:425]: Ah that makes sense, we don't support i386 from core20+. I'll remove them

File: tests/smoke/install/task.yaml (hunk 1 above)
Comment (mvo5) [reply tests/smoke/install/task.yaml:49]: Haha - nice. shellcheck is the best. It's fine if it suggested it. It's just a bit usual I gues

File: tests/lib/external/snapd-testing-tools/tools/repack-kernel (hunk 5 above)
Comment (Meulengracht) [reply tests/lib/external/snapd-testing-tools/tools/repack-kernel]: I removed the subshell here

**Pull request merged**